### PR TITLE
Improve rate-limiting windowing for the default cache backend.

### PR DESCRIPTION
**Pull request by skorokithakis**
This is related to #20. Instead of renewing the rate limit every request, this PR resets the limits every `period` seconds.

This means that:

* Originally, if you set a limit of 10/minute and performed one request every 30 seconds, you would get rate-limited after 11 requests, since the counter would never reset.
* After this change, the counter resets every minute, so performing one request every 30 seconds will never get you rate-limited.

There is an issue in this PR that will require the cache to be cleared when upgrading, due to the changing data format in the cache key. I can check the data type and upgrade backwards-compatibly, if needed.

### CODE REVIEW

File: brake/backends/cachebe.py [hunk 1]
```diff
@@ -71,7 +73,7 @@ def limit(self, func_name, request,
             if ':ip:' in counter:
                 ratelimited_by = 'ip'
 
-            if counters[counter] > count:
+            if counters[counter][0] > count:
```
Comment (gmcquillan) [inline brake/backends/cachebe.py]: May seem paranoid, but I'd guard this being an empty list.

Comment (skorokithakis) [inline brake/backends/cachebe.py]: How come? You weren't guarding against it being an empty list before...

File: brake/backends/cachebe.py [hunk 2]
```diff
@@ -51,12 +52,13 @@ def _keys(self, func_name, request, ip=True, field=None, period=None):
...
+        counters = dict((key, (1, time.time() + period)) for key in self._keys(
             func_name, request, ip, field, period))
         counters.update(cache.get_many(counters.keys()))
-        for key in counters:
-            counters[key] += 1
-        cache.set_many(counters, timeout=period)
+        for key, value in counters.items():
+            count, expiration = value
+            count += 1
+            cache.set(key, (count, expiration), timeout=expiration - time.time())
```
Comment (gmcquillan) [inline brake/backends/cachebe.py]: Okay, this is pretty clever. So when you pass a tuple into the value for ```cache.set``` how does it serialize, pickle?

Also, am I missing something, or is nothing reading the expiration out of the memcached value?

File: brake/backends/cachebe.py (hunk 1 above)
Comment (gmcquillan) [inline brake/backends/cachebe.py]: Before it was a scalar, now it's a tuple.

Not super worried about it, but this needs some careful testing.

File: brake/backends/cachebe.py (hunk 2 above)
Comment (skorokithakis) [inline brake/backends/cachebe.py]: > So when you pass a tuple into the value for cache.set how does it serialize, pickle?

I'm not sure, I just let it use the default Django cache serialization. I can check to see what the cache serializer is.

> am I missing something, or is nothing reading the expiration out of the memcached value?

What do you mean? The value is read in `cache.get_many()` in line 56 and reused to set the timeout again.

File: brake/backends/cachebe.py (hunk 1 above)
Comment (skorokithakis) [inline brake/backends/cachebe.py]: True, I can add it if you want.

File: brake/backends/cachebe.py (hunk 2 above)
Comment (gmcquillan) [inline brake/backends/cachebe.py]: Ah, I see. It's hard to follow what's happening here, because it's assigned implicitly on line 55. 

Can you add some tests that make sure that these conditions are invariant?

I'm especially wondering what happens on the expiration boundary. Say the distance between your expiration and time.time in line 61 is greater than the period duration? We'd have a negative timeout.

Although, it looks like memcached would just set the expiration in the recent past (and it would therefore be expired): https://github.com/memcached/memcached/blob/master/memcached.c#L169

Seriously, though, this logic change needs testing.

File: brake/backends/cachebe.py (hunk 1 above)
Comment (gmcquillan) [inline brake/backends/cachebe.py]: If we can prove in testing that there's always a value there, then I'm good.

**Comment by gmcquillan** [comment]
Looks good, I'll merge it when tests are up to date and passing. I'll take on updating docs unless you'd prefer to. I think they should be converted to MD, anyway.

File: brake/backends/cachebe.py (hunk 2 above)
Comment (skorokithakis) [inline brake/backends/cachebe.py]: I can test it, but Travis isn't liking the tests on this branch, for some reason. It can't import unittest, even though there's nothing in this PR that would make that fail.

File: brake/backends/cachebe.py (hunk 1 above)
Comment (skorokithakis) [inline brake/backends/cachebe.py]: I'm not very clear on what you're wary of, I'm afraid. This is always going to have a value, and it's going to be the value that was set for the cache earlier on. If it's not, the way it was before would break as well. I'm not seeing why a tuple is more "dangerous" than an int.

Comment (gmcquillan) [inline brake/backends/cachebe.py]: It's not a big deal, but we're moving from a scalar value that is referred to by its name, to an opaque repeated data structure and blindly grabbing whatever is in its first index. 

Now, after talking to you, I know what that is. But this these are called "magic numbers" for a reason, and that's because they omit context.

It's a nit of a comment, if  you really feel strongly about it that's fine. I don't feel strongly; I just thought I'd bring it up.

File: brake/backends/cachebe.py (hunk 2 above)
Comment (gmcquillan) [inline brake/backends/cachebe.py]: Sure, just make sure it's passing locally. I'll deal with Travis.

File: brake/backends/cachebe.py (hunk 1 above)
Comment (skorokithakis) [inline brake/backends/cachebe.py]: I don't really feel strongly about it, I'm just trying to understand what you're wary about. I think we should check regardless and replace the value if it's not a tuple, as that means that the library was just upgraded and there is old, integer-only data in the cache. I'll commit that soon.

**Comment by skorokithakis** [comment]
Alright, tests are passing!

File: brake/backends/cachebe.py [hunk 3]
```diff
@@ -71,7 +78,11 @@ def limit(self, func_name, request,
             if ':ip:' in counter:
                 ratelimited_by = 'ip'
 
-            if counters[counter] > count:
+            current_count = counters[counter]
+            if isinstance(current_count, tuple):
```
Comment (gmcquillan) [inline brake/backends/cachebe.py]: This is nice, I think it gracefully handles the case where you're upgrading from the existing cache system. However, I realize there's still a problem for existing (old) django installations that are expecting a scalar value. (i.e. since you're not going to redeploy all of your web front ends at precisely the same time, there's going to be a window for the ratelimit code to throw exceptions on the older versions.

The way I see it, we have two options: call this a major version upgrade and let clients deal with the fallout, or we can release an intermediate version that doesn't write your new value, but *does* check for a tuple value and continues as normal. Then people can upgrade to the intermediate version with no race conditions in the value, and upgrade to this, enhanced version with no troubles.

I kind of like the second option best.

@skorokithakis what do you think?

Comment (skorokithakis) [inline brake/backends/cachebe.py]: That sounds reasonable to me, but it means that we'll have to wait for a whole version on the old ratelimiting function, which turned out to be pretty bad for our users. Since there's always a window of possible race conditions when upgrading, I'd be in favor of making this a major release and warning people about the possible breakage.

Then again, my upgrades usually take 10 seconds to upgrade every machine, and we're fine with a 10 second window. I don't know if people upgrade things with days between each machine, so I guess it depends on how many people will be impacted.

Comment (gmcquillan) [inline brake/backends/cachebe.py]: Yeah, it's a whole upgrade cycle, but you can skip it if you want to. It provides people who are more sensitive to faults in their systems a chance to avoid that. I'll make a branch that pulls in that logic and try to cut a release before merging this. 

I'll start on that so we can still get this out there today.

**Comment by gmcquillan** [comment]
You can merge master into your branch now and Travis will work.

**Comment by gmcquillan** [comment]
Future proofing release 1.4.1 has been released. I'll get these changes merged tomorrow, and provide some upgrade instructions, and you should be good to go.

**Comment by skorokithakis** [comment]
Are there any upgrade instructions on this? We're hitting this issue with infinitely-blocked clients and would really like to use the fixed version.

**Comment by gmcquillan** [comment]
Working on it. I've been out sick for the past few days, so I haven't really had any free time. I should actually be able to merge your changes in, add some docs and release today.

**Comment by gmcquillan** [comment]
See PR #24. Merged Changes into master there, added some documentation.

**Comment by gmcquillan** [comment]
@skorokithakis Please review the other PR, and I'll go ahead and publish the latest version after merging.
